roachtest/sqlsmith: non-nullable column with no value

roachtest.sqlsmith/setup=empty/setting=default [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3035007&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3035007&tab=artifacts#/sqlsmith/setup=empty/setting=default) on master @ [7762f00fdb5dd7a65545b684fbd4f0babe383a76](https://github.com/cockroachdb/cockroach/commits/7762f00fdb5dd7a65545b684fbd4f0babe383a76):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=default/run_1
	sqlsmith.go:215,sqlsmith.go:251,test_runner.go:801: error: pq: internal error: Non-nullable column "tab_1699:col0_8" with no value! Index scanned was "primary" with the index key columns (col0_5,col0_7,col0_0,col0_4) and the values (0.00989532636966406,526710524,'BOX(-1.1538628222617553 -0.6825724451748342,0.8687846229949832 -0.6060055664340925)',0.6612314127415079736)
		stmt:
		ALTER TABLE defaultdb.public.tab_1699
			ADD COLUMN
				col_3533
					INTERVAL
					DEFAULT (
						(0.05654613720162893:::FLOAT8::FLOAT8 * '-43 years -4 mons -647 days -20:22:13.181403':::INTERVAL::INTERVAL)::INTERVAL
					);
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=empty/setting=default
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=default.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/backfill/backfill.go
+++ b/pkg/sql/backfill/backfill.go
@@ -108,9 +108,13 @@ func (cb *ColumnBackfiller) init(
 		cb.updateExprs[j+len(cb.added)] = tree.DNull
 	}
 
-	// We need all the columns.
+	// We need all the non-virtual columns.
 	var valNeededForCol util.FastIntSet
-	valNeededForCol.AddRange(0, len(desc.PublicColumns())-1)
+	for i, c := range desc.PublicColumns() {
+		if !c.IsVirtual() {
+			valNeededForCol.Add(i)
+		}
+	}
 
 	tableArgs := row.FetcherTableArgs{
 		Desc:            desc,

--- a/pkg/sql/logictest/testdata/logic_test/virtual_columns
+++ b/pkg/sql/logictest/testdata/logic_test/virtual_columns
@@ -1109,3 +1109,21 @@ CREATE TABLE public.t63167_b (
    CONSTRAINT "primary" PRIMARY KEY (rowid ASC),
    FAMILY "primary" (a, rowid)
 )
+
+# Test that columns backfills to tables with virtual columns work.
+subtest column_backfill
+
+statement ok
+CREATE TABLE t_65915 (i INT PRIMARY KEY, j INT AS (i + 1) VIRTUAL NOT NULL);
+INSERT INTO t_65915 VALUES (1)
+
+statement ok
+ALTER TABLE t_65915 ADD COLUMN k INT DEFAULT 42;
+
+query III
+SELECT * FROM t_65915;
+----
+1  2  42
+
+statement ok
+DROP TABLE t_65915

